PR: Updated BeamMP-Chat + UI

CarlFox98 opened the pull request:
I did the following work(AI tools like GitHub CoPilot and ChatGPT 3.5 were also used):

1) Modified and adjusted app.html so the BeamMP-Chat UI app looks much nicer (IMO).

2) Modified and adjusted app.json with more up to date information.

3) Modified and adjusted app.js to displayer the user's time better using locales and time zone. Also adjusted some of the other functions and added more descriptive comments.

4) Made a new app.png in GIMP.

5) Created an app.css file as a redundant backup in case of emergencies. (Does nothing and is empty except for some comments.)

NeotericGamer98 committed:
Updated app.html for BeamMP-Chat

Still needs further testing and work.

NeotericGamer98 committed:
Modified BeamMP-Chat + UI

I did the following work:

1) Modified and adjusted app.html so the BeamMP-Chat UI app looks much nicer (IMO).

2) Modified and adjusted app.json with more up to date information.

3) Modified and adjusted app.js to displayer the user's time better using locales and time zone. Also adjusted some of the other functions and added more descriptive comments.

4) Made a new app.png in GIMP.

5) Created an app.css file as a redundant backup in case of emergencies. (Does nothing and is empty except for some comments.)

NeotericGamer98 commented:
I am open to feedback or suggestions.

OfficialLambdax reviewed (changes requested):
1. The chat does not autoscrol
2. The buttons to the right of the chat input field dont have a function
3. On ui reload the app cannot reload the chat history, stating the following error 
![hjgghjjghghjjhgkkjhhjk](https://github.com/BeamMP/BeamMP/assets/91626975/1ee20b99-6516-4ca2-8594-c6799a03e393)

NeotericGamer98 commented:
> 1. The chat does not autoscrol
> 
> 2. The buttons to the right of the chat input field dont have a function
> 
> 3. On ui reload the app cannot reload the chat history, stating the following error 
> 
> ![hjgghjjghghjjhgkkjhhjk](https://github.com/BeamMP/BeamMP/assets/91626975/1ee20b99-6516-4ca2-8594-c6799a03e393)
> 
> 

I'll look into fixing these issues. 

Thank you for letting me know.

Lucky4Luuk commented:
Would it be possible to get a little preview here of what the chat looks like? I think considering this PR is talking about making it look better, it'd be nice to at least know what it's going to look like :)

OfficialLambdax commented:
> Would it be possible to get a little preview here of what the chat looks like? I think considering this PR is talking about making it look better, it'd be nice to at least know what it's going to look like :)

from the discord
![kkkjkkjk](https://github.com/BeamMP/BeamMP/assets/91626975/7a8e4064-05a6-4170-8ed5-954f18fc135a)
![jhkljhjkjklkljlk](https://github.com/BeamMP/BeamMP/assets/91626975/f34efbf1-8354-430f-b5b6-2195b0f19ea3)

Lucky4Luuk commented:
Oh, I really don't like that, sorry. There's so much dead space and the margins are so massive. In my opinion, a chat UI should be compact, support colors and hyperlinks and that's it. Simple to read, easy to ignore and not massive. Perfect example is the Minecraft chat, it does exactly what it needs to do. I would really prefer this doesn't get merged, the current chat might not be the greatest but it's more functional than this.

NeotericGamer98 commented:
> Oh, I really don't like that, sorry. There's so much dead space and the margins are so massive. In my opinion, a chat UI should be compact, support colors and hyperlinks and that's it. Simple to read, easy to ignore and not massive. Perfect example is the Minecraft chat, it does exactly what it needs to do. I would really prefer this doesn't get merged, the current chat might not be the greatest but it's more functional than this.

I agree, I'll work on making the chat more streamlined and compact. I'll also try and resolve the bugs that were mentioned. 

However I'd like to keep the Discord styled color scheme if that's okay.

Lucky4Luuk commented:
Does BeamMP not already have its own black/orange colour scheme? Might be better to stick to that, it makes more sense than using the Discord colour scheme.

lionkor commented:
I would definitely reduce the left- and right margins there, these are a bit jarring 
![image](https://github.com/BeamMP/BeamMP/assets/29932116/28892064-08b9-4399-82e4-e5c87d98e1fe)

So that they align on both sides.

I would also suggest to move the button to the right. The final result could be something like:

![image](https://github.com/BeamMP/BeamMP/assets/29932116/fb463b7b-0d23-49d4-9ba5-544c0eba7d93)

of course with some of the in-chat-bubble-margins changed or something.

lionkor commented:
I woul also left-align the text inside the "chat here" box.

lionkor commented:
And the send button should be bigger, or completely disappear. Maybe like this size (ignore text size/stretch)
![image](https://github.com/BeamMP/BeamMP/assets/29932116/6457621e-4ce1-4044-baaf-9c804e708cba)

Starystars67 commented:
I feel like this tries to be a little too much like discord or something. Though the chat does need a revamp.

Pretty sure someone also mentioned about possibly using vue for this so that may be an option too. Not sure just yet of what benefits this may or may not offer though.

NeotericGamer98 commented:
> Does BeamMP not already have its own black/orange colour scheme? Might be better to stick to that, it makes more sense than using the Discord colour scheme.

I can experiment with a theme like that.

NeotericGamer98 commented:
> I would definitely reduce the left- and right margins there, these are a bit jarring ![image](https://private-user-images.githubusercontent.com/29932116/285603144-28892064-08b9-4399-82e4-e5c87d98e1fe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDA5MzEyNzEsIm5iZiI6MTcwMDkzMDk3MSwicGF0aCI6Ii8yOTkzMjExNi8yODU2MDMxNDQtMjg4OTIwNjQtMDhiOS00Mzk5LTgyZTQtZTVjODdkOThlMWZlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI1VDE2NDkzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNkYTk1YzMwNTc5OGM3YTJlMmU0ZWU2MDBmMGE1YjJmYmI2ZDExOGI4M2FmMmRiMzZmMTgxYTAxMmE4YWU0M2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.4pfMTJxpgiQTC_deilKIAKeTNb4DX7klpGqfHryhrzw)
> 
> So that they align on both sides.
> 
> I would also suggest to move the button to the right. The final result could be something like:
> 
> ![image](https://private-user-images.githubusercontent.com/29932116/285603352-fb463b7b-0d23-49d4-9ba5-544c0eba7d93.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDA5MzEyNzEsIm5iZiI6MTcwMDkzMDk3MSwicGF0aCI6Ii8yOTkzMjExNi8yODU2MDMzNTItZmI0NjNiN2ItMGQyMy00OWQ0LTliYTUtNTQ0YzBlYmE3ZDkzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI1VDE2NDkzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgxYTZhYjBlZjM1NDJlOTgxZTViMThjZDRhYzUxZWYyMjk0NzlmNDQyODYxM2VkZjdlNzlhZjBmYzk1M2QzMTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.xzxaeJDPgyZGxhtCn8Wz28arDWu5hSrIcKRLeZnHs9s)
> 
> of course with some of the in-chat-bubble-margins changed or something.

I will play around with the margin and padding values and see what I can do to reduce the "bloat".

NeotericGamer98 commented:
> I feel like this tries to be a little too much like discord or something. Though the chat does need a revamp.
> 
> Pretty sure someone also mentioned about possibly using vue for this so that may be an option too. Not sure just yet of what benefits this may or may not offer though.

Noted, I will try and make it less Discord like then.

OfficialLambdax commented:
Yeah people have preferences. If you want to keep your design, A tiny button in the chat ui that switches between the different designs could solve that

Lucky4Luuk commented:
Why not different UI apps? That functionality is already built into beam, I think it is nicer than just adding a button to toggle styles. Perhaps this way, people will also be inspired to make their own chat UI apps?

NeotericGamer98 committed:
Updated BeamMP-Chat

Updated the app.html styling to NOT look like Discord. I did my best to make it look as unobtrusive as possible.

Updated app.js and refined a couple of functions. Auto-Scroll and swap functions seem to work fine.

Updated app.json slightly.

Added app.css as a backup file.

NeotericGamer98 commented:
Just pushed a new commit. 

Make sure when you're testing and peer reviewing to use the correct app.
For some reason by default when loading into a server it seems to loaded a version of the app with not all the styling applied to it. 

To properly test my changes to the app do the following while in game:

1) Hit ESC to open up the BeamNG menu.
2) Click on APPS on the top menu.
3) Scroll down until you see BeamMP Chat (Dark Grey background with BeamMP logo on it)
4) Click add.
5) Remove the old chat app as to avoid confusion.
6) Happy testing.

Starystars67 commented:
Hey Neo,

Not got the opportunity to test as yet but could you include a screenshot of your final design so we can all see it here without needing to load up the game?

Thanks

Starystars67 commented on ui/modules/apps/BeamMP-Chat/app.json:
```diff
@@ -1,17 +1,17 @@
+    "types": [
+        "ui.apps.categories.multiplayer"
+    ],
+    "version": "1.0"
```
Given this is a newer version, please increase this to 1.1.

CarlFox98 replied:
Will do.

CarlFox98 replied:
Resolved.

Starystars67 commented on ui/modules/apps/BeamMP-Chat/app.css:
What is the point of this file?
Git's version control allows us to look back at older versions if we need to so I am not sure why this is needed unless it is used?

CarlFox98 replied:
To act as a backup, just in case.

Starystars67 commented on ui/modules/apps/BeamMP-Chat/app.css:
```diff
@@ -0,0 +1,231 @@
+/* BeamNG Multiplayer Chat Styling */
```
This file is also not BeamNG's so please correct the wording here.
Something like `BeamMP Multiplayer Chat Styling` is fine.

CarlFox98 replied:
Will do, sorry for the typo.

CarlFox98 replied:
Issue resolved.

Starystars67 commented on ui/modules/apps/BeamMP-Chat/app.html:
Again is the CSS external or not?

CarlFox98 replied:
External CSS linking does not work with the app. 

I tried to do that before and it failed to load properly.

Starystars67 commented on ui/modules/apps/BeamMP-Chat/app.js:
```diff
@@ -62,57 +67,104 @@ app.controller("Chat", ['$scope', 'Settings', function ($scope, Settings) {
+	// Function to handle Enter key press
+	function onEnterKeyPress(event) {
+		const chatInput = document.getElementById("chat-input");
+		if (event.key === "Enter" && chatInput === document.activeElement) {
```
I wonder, could a similar principle be used to add the feature of when you press Enter it opens the chat box ready to type like most games too?

CarlFox98 replied:
Possibly, if that is something you want implemented.

OfficialLambdax replied:
Take a look at this file that beammp already containes
https://github.com/BeamMP/BeamMP/blob/development/lua/ge/extensions/core/input/actions/beammp.json

i think this way you can catch keypresses that are defined in the game controls (dunno if you can set a default binding like the enter key tho). But if so, it looks like that you can send a signal to this app.

CarlFox98 replied:
> Take a look at this file that beammp already containes https://github.com/BeamMP/BeamMP/blob/development/lua/ge/extensions/core/input/actions/beammp.json
> 
> i think this way you can catch keypresses that are defined in the game controls (dunno if you can set a default binding like the enter key tho). But if so, it looks like that you can send a signal to this app.

What would be the reason for trying to implement this into the chat app? 

Is the current method of mousing over, clicking on the chat input, typing a message, and hitting the enter key really that much of a chore? Personally, I think the current method of communication seems adequate enough. I don't really see the need in adding more possible complexity to this chat app. 

However, I am curious what the other contributors/developers think.

StanleyDudek replied:
You would set the following keybind in the BeamNG Controls, and when you pressed `Enter` you would be focused to the text input, so you can immediately type.

![image](https://github.com/BeamMP/BeamMP/assets/49531350/e7f4f1b6-69e7-4fa3-85e2-05fc385cc13f)

I think this is something that would be very welcome.

CarlFox98 replied:
> You would set the following keybind in the BeamNG Controls, and when you pressed `Enter` you would be focused to the text input, so you can immediately type.
> 
> 
> 
> ![image](https://github.com/BeamMP/BeamMP/assets/49531350/e7f4f1b6-69e7-4fa3-85e2-05fc385cc13f)
> 
> 
> 
> I think this is something that would be very welcome.

What would be involved in trying to implement this? 

Would this involve modifying files that do not reside within the BeamMP-Chat folder? 

This is something I could look into trying to add. That is, if this is what the majority of the BeamMP project developers would like.

Starystars67 replied:
Indeed this is what I meant with:

> I wonder, could a similar principle be used to add the feature of when you press Enter it opens the chat box ready to type like most games too?

NeotericGamer98 commented:
> Hey Neo,
> 
> Not got the opportunity to test as yet but could you include a screenshot of your final design so we can all see it here without needing to load up the game?
> 
> Thanks

![image](https://github.com/BeamMP/BeamMP/assets/74558635/18db7c97-3763-4ad0-a74d-42856ac74386)

NeotericGamer98 committed:
Updated BeamMP Chat

Updated app.json app version to 1.1

Updated app.css AND app.html to say BeamMP instead of BeamNG. (Accidental past typos.)

NeotericGamer98 commented:
> 1. The chat does not autoscrol
> 
>     2. The buttons to the right of the chat input field dont have a function
> 
>     3. On ui reload the app cannot reload the chat history, stating the following error
>        ![hjgghjjghghjjhgkkjhhjk](https://private-user-images.githubusercontent.com/91626975/285576547-1ee20b99-6516-4ca2-8594-c6799a03e393.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEyOTMzMzcsIm5iZiI6MTcwMTI5MzAzNywicGF0aCI6Ii85MTYyNjk3NS8yODU1NzY1NDctMWVlMjBiOTktNjUxNi00Y2EyLTg1OTQtYzY3OTlhMDNlMzkzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI5VDIxMjM1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNkOWUzYWY4MzAyODdhMjU3MGVjZTdiMjY1ZjJmMzI4YThiODgxMzg5OTllOTA5YjlhMGQyMjk4YzQ4YzZkMjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.mKItp34oeVnNcXkOiedQ5jfvUKaxiEErSPTSbyxqptA)

These issues, to the best of my knowledge, have been resolved in the newest commit.

NeotericGamer98 commented:
> I would definitely reduce the left- and right margins there, these are a bit jarring ![image](https://private-user-images.githubusercontent.com/29932116/285603144-28892064-08b9-4399-82e4-e5c87d98e1fe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEyOTMzMzcsIm5iZiI6MTcwMTI5MzAzNywicGF0aCI6Ii8yOTkzMjExNi8yODU2MDMxNDQtMjg4OTIwNjQtMDhiOS00Mzk5LTgyZTQtZTVjODdkOThlMWZlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI5VDIxMjM1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE5MjVhNDMzN2U4MGIwNDJjYjU3ZjEwZWJhNjNkNzExZWE5ZTFjMzkwYWEyYzU1Yjc4Mjc0NjZjYmU0YWU5MzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ODP02BRiEjJ0PJTw7YmliDWDIJOVkfA0I7w2vgmkg2k)
> 
> So that they align on both sides.
> 
> I would also suggest to move the button to the right. The final result could be something like:
> 
> ![image](https://private-user-images.githubusercontent.com/29932116/285603352-fb463b7b-0d23-49d4-9ba5-544c0eba7d93.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEyOTMzMzcsIm5iZiI6MTcwMTI5MzAzNywicGF0aCI6Ii8yOTkzMjExNi8yODU2MDMzNTItZmI0NjNiN2ItMGQyMy00OWQ0LTliYTUtNTQ0YzBlYmE3ZDkzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI5VDIxMjM1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA3ZDY4OTdhMGM0N2E4YzM3MDUzNTZmODViYjI0OWMzY2FmNjUzMTZkZWE2ZGUyZjJiOGYzMmEwOTVmZTY4ZTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.yFsFGF518i2WSWmen-oRI6reJteK3rdVV2TiM5V8NlE)
> 
> of course with some of the in-chat-bubble-margins changed or something.

These issues, to the best of my knowledge, have been resolved in the newest commit.

OfficialLambdax reviewed (changes requested):
The app works with all the previous functionalities. Could not get any kind of error.

However this apps design appears to be influencing other apps
![3](https://github.com/BeamMP/BeamMP/assets/91626975/88e68f9f-7bec-4490-ae62-b042b7268eb4)

Starystars67 commented:
I have moved this PR to 4.11.0 milestone due to outstanding work and uncertainty.
